### ISSUE TITLE
[Bug]: GLM 5.2 MTP didn't work in AMD MI300x

### ISSUE THREAD

**Issue by haosdent** [vllm-project/vllm#48568]
### 🐛 Describe the bug

GLM-5.2-FP8 with MTP speculative decoding hangs on 8x MI300X (gfx942). Startup finishes, but the engine deadlocks in the vocab-parallel logits all-gather on the first real spec-decode step (RCCL `no transport for peer on channel N`).

**Repro:** `zai-org/GLM-5.2-FP8`, ROCm+AITER, `--tensor-parallel-size 8 --kv-cache-dtype fp8_e4m3 --linear-backend aiter --moe-backend aiter --speculative-config '{"method":"mtp","num_speculative_tokens":K}'`.

**Root cause 1 (hits the config above):** warmup runs `compute_logits`' all-gather at 1 row/req, but spec decode verifies `1 + num_speculative_tokens` rows/req. That larger message maps onto extra RCCL channels whose transports connect lazily on first use, and that first use is a live decode step, so the connect fails and hangs.
Fix: after the profile `_dummy_run`, warm `compute_logits` once at `min(max_num_reqs*(1+num_spec_tokens), max_num_tokens)` rows (guarded by `num_spec_tokens > 0`).

**Root cause 2 (only with `--enable-expert-parallel`, TP>1):** under sequence-parallel MoE, MTP's `mtp_block` reduce-scatters its output but `DeepSeekMultiTokenPredictorLayer.forward` never restores full tokens (unlike `DeepseekV2Model.forward`), so ranks feed mismatched token counts into the collective. Same class as the Qwen MTP fix #48429.
Fix: when `hidden_states.shape[0] != positions.shape[0]`, all-gather `[hidden_states, residual]`, slice to full length, and split.

Fix 1 validated e2e on MI300X (coherent output, num_spec=1 accept 75-82%).

**Comment by vllmellm** [comment]
Hi, I've tested in the `vllm/vllm-openai-rocm:v0.25.1` image, with this command, and serving is working fine. Which vllm version are you running on? 

```
# I ran into VRAM OOM issue with full context length
VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1 vllm serve zai-org/GLM-5.2-FP8 \
  --kv-cache-dtype fp8_e4m3 \
  --tensor-parallel-size 8 \
  --linear-backend aiter \
  --moe-backend aiter \
  --tool-call-parser glm47 \
  --enable-auto-tool-choice \
  --reasoning-parser glm45 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":5}' --port 8030 --max-model-len 256000 --gpu-memory-utilization 0.7
```

At a single request:
```
(APIServer pid=13798) INFO 07-16 06:44:27 [loggers.py:273] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 45.4 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.6%
(APIServer pid=13798) INFO 07-16 06:44:27 [metrics.py:120] SpecDecoding metrics: Mean acceptance length: 3.53, Accepted throughput: 32.60 tokens/s, Drafted throughput: 64.49 tokens/s, Accepted: 326 tokens, Drafted: 645 tokens, Per-position acceptance rate: 0.853, 0.643, 0.496, 0.318, 0.217, Avg Draft acceptance rate: 50.5%
```

**Comment by haosdent** [comment]
@vllmellm I forgot the exact error, but even with the latest main, I still need to apply #48572 to make it work.

**Comment by vllmellm** [comment]
@haosdent thanks for the reply. Can you test it with docker container? or how did you setup your local environment?

I've tried using this image `vllm/vllm-openai-rocm:nightly-6f91edf96d3f3272945809c04702380053bff4de` which is on this commit `6f91edf96`  and the MTP is working with the same command I tested on the `vllm/vllm-openai-rocm:v0.25.1` image. 

```
# I ran into VRAM OOM issue with full context length
VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1 vllm serve zai-org/GLM-5.2-FP8 \
  --kv-cache-dtype fp8_e4m3 \
  --tensor-parallel-size 8 \
  --linear-backend aiter \
  --moe-backend aiter \
  --tool-call-parser glm47 \
  --enable-auto-tool-choice \
  --reasoning-parser glm45 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":5}' --port 8030 --max-model-len 256000 --gpu-memory-utilization 0.7
```

**Comment by haosdent** [comment]
Thanks for testing. Two reasons it doesn't reproduce there, and why the fix is still needed.

**Root cause 2** is out of scope for your command: it only triggers under sequence-parallel MoE, and you didn't pass `--enable-expert-parallel`.

**Root cause 1** is a lazy-connect race, so not reproducing is expected in many setups. Whether the larger spec-decode message lands on a not-yet-connected RCCL channel depends on channel count and topology. We confirmed the mechanism directly: both `NCCL_MIN_NCHANNELS=32` (force-connect) and `NCCL_MAX_NCHANNELS=8` (cap below the channel involved) make it coherent. Neither is production-viable for us: forcing 32 fragments the decode all-reduce to ~30 tok/s, capping at 8 pushes TTFT to 25-38s. Hence the warmup fix.

The gap itself is real on current main and not ROCm-specific: `_dummy_run` builds `logit_indices = np.cumsum(num_scheduled_tokens) - 1`, one row per request, so `compute_logits` is only ever warmed at `num_reqs` rows, while real spec decode runs it at `num_reqs * (1 + num_spec_tokens)`. None of the four `compute_logits` call sites warms the spec-sized shape during init, and it sits outside the cudagraph so capture doesn't cover it.

Side effect worth noting: `profile_run()` is what `determine_available_memory` measures, so today the `num_reqs * num_spec_tokens * vocab_size` logits allocation never shows up in the peak. The patch makes the profile match what decode actually allocates, on any backend.

FWIW we run this in production on 8x MI300X, `num_speculative_tokens=3`, 256k context at `--gpu-memory-utilization 0.85`, per-position acceptance 0.93 / 0.85 / 0.74 (mean acceptance length 3.3-3.6).

**Comment by vllmellm** [comment]
Thanks for you the explanation. On that same image `vllm/vllm-openai-rocm:nightly-6f91edf96d3f3272945809c04702380053bff4de` I ran the following again with `--enable-expert-parallel` and the EP is working. Can you test with in this image? 

```
# I ran into VRAM OOM issue with full context length
VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1 vllm serve zai-org/GLM-5.2-FP8 \
  --kv-cache-dtype fp8_e4m3 \
  --tensor-parallel-size 8 \
  --enable-expert-parallel \
  --linear-backend aiter \
  --moe-backend aiter \
  --tool-call-parser glm47 \
  --enable-auto-tool-choice \
  --reasoning-parser glm45 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":5}' --port 8030 --max-model-len 256000 --gpu-memory-utilization 0.7

# server logs
...
(Worker_TP0_EP0 pid=13332) INFO 07-30 03:46:22 [expert_map_manager.py:245] [EP Rank 0/8] Expert parallelism is enabled. Expert placement strategy: linear. Local/global number of experts: 33/256. Experts local to global index map: 0->0, 1->1, 2->2, 3->3, 4->4, 5->5, 6->6, 7->7, 8-
>8, 9->9, 10->10, 11->11, 12->12, 13->13, 14->14, 15->15, 16->16, 17->17, 18->18, 19->19, 20->20, 21->21, 22->22, 23->23, 24->24, 25->25, 26->26, 27->27, 28->28, 29->29, 30->30, 31->31, 32->256.
...
(APIServer pid=12801) INFO 07-30 05:05:49 [loggers.py:310] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 82.3 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 46.1%
(APIServer pid=12801) INFO 07-30 05:05:49 [metrics.py:120] SpecDecoding metrics: Mean acceptance length: 3.12, Accepted throughput: 56.00 tokens/s, Drafted throughput: 132.00 tokens/s, Accepted: 560 tokens, Drafted: 1320 tokens, Per-position acceptance rate: 0.830, 0.557, 0.375, 0.212, 0.148, Avg Draft acceptance rate: 42.4%
```

**Comment by tianxiaojiang4** [comment]
Possibly related: #52833. Same model on gfx950 (MI355X), vLLM 0.27.1+rocm723 — no hang there, but MTP silently accepts 0% of drafts under `--enable-expert-parallel` with TP8, and dropping EP dies with `hipErrorIllegalAddress` during memory profiling instead. Ruled out draft chaining (`num_speculative_tokens=1` is also 0.000 at position 0) and the quark exclusions on the unquantized nextn layer.

(Edited: I originally cited an SGLang run as a working reference. That run pins acceptance via `SGLANG_SIMULATE_ACC_LEN`, so it's simulated, not measured — withdrawing that comparison. The vLLM numbers above are all measured.)
